sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala

```diff
@@ -465,6 +479,11 @@ class SessionCatalog(
       tableName: TableIdentifier,
       specs: Seq[TablePartitionSpec],
       newSpecs: Seq[TablePartitionSpec]): Unit = {
+    val tableMetadata = getTableMetadata(tableName)
+    // Verify if old and new specs exactly match the partition spec of the table definition
+    Seq(specs, newSpecs).flatten.foreach { s =>
+      requireExactMatchedPartitionSpec(s.keys.toSeq, tableMetadata.partitionColumnNames, tableName)
+    }
```

Comment:
if you take my suggestion then this would become:
```
requirePartitionsDefinedInTables(specs, tableMetadata)
requirePartitionsDefinedInTables(newSpecs, tableMetadata)
```
a little simpler I think